Generate passwords.

psbleep opened the pull request:
Partially addresses #14.

psbleep committed:
Adding feature: generate new passwords.

Kwpolska commented on upass/__main__.py:
```diff
@@ -274,6 +279,70 @@ def copy_selected(self, originator):
+            path = self.current + '/'
+        self.set_header('GENERATE PASSWORD')
+        self._clear_box()
+        self.generate_password_path = urwid.Edit(("highlight", "{0}{1}".format(GENERATE_PASSWORD_PROMPT, path)))
```
I would prefer to put the path as the default text, so it can be edited by the user in case they need it.

```py
self.generate_password_path = urwid.Edit(("highlight", "Password Name: "), path)
```

(This will also help get rid of the constant. BTW, the `+` operator would look better in this case.)

Kwpolska commented on upass/__main__.py:
```diff
@@ -274,6 +279,70 @@ def copy_selected(self, originator):
+        self._clear_box()
+        self.generate_password_path = urwid.Edit(("highlight", "{0}{1}".format(GENERATE_PASSWORD_PROMPT, path)))
+        self.generate_password_length = urwid.IntEdit("Length: ", default=length)
+        self.generate_password_symbols = urwid.CheckBox("Symbols? ", state=symbols)
```
[text style] Remove the question mark and trailing space here and below.

Kwpolska commented on upass/__main__.py:
```diff
@@ -274,6 +279,70 @@ def copy_selected(self, originator):
+        self.loop.draw_screen()
+        p = subprocess.Popen(gargs, stdout=subprocess.PIPE,
+                             stderr=subprocess.PIPE)
+        if path.endswith('/'):
```
Make the check first, then run `Popen`

Kwpolska commented on upass/__main__.py:
```diff
@@ -274,6 +279,70 @@ def copy_selected(self, originator):
+            errormsg = 'Invalid key path "{0}". Please enter a valid password name.'.format(path)
+        else:
+            try:
+                stdout, errormsg = p.communicate(timeout=1)
```
The timeout is too tight and might fail on some systems (it does on mine, for example). Two better solutions:

1. use `p.stdout` / `p.stderr` instead, and check the pass output (by text comparison)
2. use `os.path.exists(path + '.gpg')`

Kwpolska commented on upass/__main__.py:
```diff
@@ -274,6 +279,70 @@ def copy_selected(self, originator):
+                errormsg = ('Password with name "{0}" already exists. '
+                            'Please select the option to overwrite the existing password to generate a new password '
+                            'for this entry.'.format(path))
+        if errormsg:
```
The app hangs when displaying the error screen for some reason.

Kwpolska reviewed (changes requested):
That’s a good idea, but there are a few glitches in it.

psbleep committed:
Cleaning up password generating feature.

Checks for problems with the key path name or already-existing
keys before calling the "pass generate" command, instead of after
calling it.

Checks for already-existing keys by looking on the file system, rather
than timing out.

Error screen does not hang (at least on my machine).

Cleans up style issues.

psbleep commented:
Thanks for reviewing and giving feedback! The new commit should fix the problems you noticed.

Kwpolska commented on upass/__main__.py:
```diff
@@ -274,6 +276,66 @@ def copy_selected(self, originator):
+            path = self.current + '/'
+        self.set_header('GENERATE PASSWORD')
+        self._clear_box()
+        self.generate_password_path = urwid.Edit(("highlight", "Passowrd Name: "), path)
```
Typo.

Kwpolska commented on upass/__main__.py:
```diff
@@ -274,6 +276,66 @@ def copy_selected(self, originator):
+            gargs.append('-n')
+        if force:
+            gargs.append('-f')
+        path = self.generate_password_path.text[len("Password Name: "):]
```
```py
path = self.generate_password_path.edit_text
```

Kwpolska commented on upass/__main__.py:
```diff
@@ -274,6 +276,66 @@ def copy_selected(self, originator):
+                        'for this entry.'.format(path))
+        else:
+            p = subprocess.Popen(gargs, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            stdout, errormsg = p.communicate()
```
a 5-second timeout would be a good idea. also, perhaps we could display the password if it’s successfully generated?

psbleep replied:
Thanks, I will put the longer timeout back in. For the second point, currently the password is loaded after being created, so the user can choose to display it from there (or copy it, if they prefer). Do you think it's better to go straight to showing the new password?

Kwpolska replied:
I think yes, so we work the same way `pass generate` does (without requiring a password).

psbleep replied:
Makes sense, will submit a new commit soon.

Kwpolska reviewed (changes requested):
Still a few things to fix…
PS. please add yourself to the `AUTHORS` file.

psbleep committed:
Displaying password after generating. Fixing modes.

Presenting the standard output from the `pass generate` command
instead of loading the new password by default.

Adding a 5 second timeout and error handling in case the
`pass generate` command hangs.

Fixing a bug where "j" and "k" keys are handled as vim-like
key-bindings on on the screen for generating a new password.
Using new modes: `generate` and `generate_results`.

Fixing a typo on the generate password screen.

Adding myself to the AUTHORS file.

Kwpolska commented:
Thanks for contributing this!

There are a few minor things I’d like to tweak here, but other than that, it’s awesome work! A new release will go out soon-ish.

psbleep commented:
Sorry, missed this notification! Thanks for helping me through my pull request, and for your work on this project.
